subsys/bluetooth/mesh/transport.c

```diff
@@ -1019,6 +1014,107 @@ int bt_mesh_ctl_send(struct bt_mesh_net_tx *tx, u8_t ctl_op, void *data,
 	return bt_mesh_net_send(tx, buf, cb, cb_data);
 }
 
+static int ctl_send_seg(struct bt_mesh_net_tx *tx, u8_t ctl_op,
+			void *data, size_t data_len, u64_t *seq_auth,
+			const struct bt_mesh_send_cb *cb, void *cb_data)
+{
+	u8_t seg_o = 0;
+	u16_t seq_zero;
+	struct seg_tx *tx_seg;
+	int i;
+	u16_t unsent = data_len;
+
+	for (tx_seg = NULL, i = 0; i < ARRAY_SIZE(seg_tx); i++) {
+		if (!seg_tx[i].nack_count) {
+			tx_seg = &seg_tx[i];
+			break;
+		}
+	}
+
+	if (!tx_seg) {
+		BT_ERR("No multi-segment message contexts available");
+		return -EBUSY;
+	}
+
+	tx_seg->dst = tx->ctx->addr;
+	tx_seg->seg_n = (data_len - 1) / 8;
+	tx_seg->nack_count = tx_seg->seg_n + 1;
+	tx_seg->seq_auth = SEQ_AUTH(BT_MESH_NET_IVI_TX, bt_mesh.seq);
+	tx_seg->sub = tx->sub;
+	tx_seg->new_key = tx->sub->kr_flag;
+	tx_seg->cb = cb;
+	tx_seg->cb_data = cb_data;
+
+	if (tx->ctx->send_ttl == BT_MESH_TTL_DEFAULT) {
+		tx_seg->ttl = bt_mesh_default_ttl_get();
+	} else {
+		tx_seg->ttl = tx->ctx->send_ttl;
+	}
+
+	seq_zero = tx_seg->seq_auth & TRANS_SEQ_ZERO_MASK;
+
+	for (seg_o = 0; seg_o <= tx_seg->seg_n; seg_o++) {
+		struct net_buf *seg;
+		u16_t len;
+		int err;
+
+		seg = bt_mesh_adv_create(BT_MESH_ADV_DATA, tx->xmit,
+					 BUF_TIMEOUT);
+		if (!seg) {
+			BT_ERR("Out of segment buffers");
+			seg_tx_reset(tx_seg);
+			return -ENOBUFS;
+		}
+
+		BT_MESH_ADV(seg)->seg.attempts = SEG_RETRANSMIT_ATTEMPTS;
+
+		net_buf_reserve(seg, BT_MESH_NET_HDR_LEN);
+		net_buf_add_u8(seg, TRANS_CTL_HDR(ctl_op, 1));
+		net_buf_add_u8(seg, (tx->aszmic << 7) | seq_zero >> 6);
+		net_buf_add_u8(seg, (((seq_zero & 0x3f) << 2) | (seg_o >> 3)));
+		net_buf_add_u8(seg, ((seg_o & 0x07) << 5) | tx_seg->seg_n);
+		len = MIN(unsent, 8);
+		net_buf_add_mem(seg, (char *)data + (data_len - unsent), len);
+		unsent -= len;
+		tx_seg->seg[seg_o] = net_buf_ref(seg);
+
+		BT_DBG("Sending %u/%u", seg_o, tx_seg->seg_n);
+
+		err = bt_mesh_net_send(tx, seg,
+				       seg_o ? &seg_sent_cb : &first_sent_cb,
+				       tx_seg);
+		if (err) {
+			BT_ERR("Sending segment failed");
+			seg_tx_reset(tx_seg);
+			return err;
+		}
+	}
+
+	if (!BT_MESH_ADDR_IS_UNICAST(tx->ctx->addr)) {
+		/* remove ack wait */
+		seg_tx_reset(tx_seg);
+	}
```

Comment:
I just realised this part is probably wrong. We don't do it for access layer messages either. IIRC even if the destination is a group address we still want the tx_seg timers and retransmissions to keep running, even though we won't receive any acks. @trond-snekvik can you confirm?

@AhmedHussein751 no need to push any changes regarding this - I'll do it myself (like I've done with the cleanups) when we reach consensus.

Reply:
You're right, we still do periodic retransmissions for group messages

Reply:
Thanks for confirming. I've pushed an update now.